connectors/memory/memory.go

```diff
@@ -0,0 +1,493 @@
+// these rows are kept ordered so that reads are lightning fast and searches are quick too
+// the row itself is a map of field name to value (map[string]dosaFieldValue])
+//
+// A read-write mutex lock is used to control concurrency, making reads work in parallel but
+// writes are not. There is no attempt to improve the concurrency of the read or write path by
+// adding more granular locks.
+type Connector struct {
+	base.Connector
+	data map[string]map[string][]map[string]dosa.FieldValue
+	lock sync.RWMutex
+}
+
+// partitionKeyBuilder extracts the partition key components from the map and encodes them,
+// generating a unique string. It uses the encoding/gob method to make a byte array as the
+// key, and returns this as a string
+func partitionKeyBuilder(ei *dosa.EntityInfo, values map[string]dosa.FieldValue) string {
+	encodedKey := bytes.Buffer{}
+	encoder := gob.NewEncoder(&encodedKey)
+	for _, k := range ei.Def.Key.PartitionKeys {
+		_ = encoder.Encode(values[k])
+	}
+	return string(encodedKey.Bytes())
+}
+
+// findInsertionPoint locates the place within a partition where the data belongs.
+// It inspects the clustering key values found in the insertMe value and figures out
+// where they go in the data slice. It doesn't change anything, but it does let you
+// know if it found an exact match or if it's just not there. When it's not there,
+// it indicates where it is supposed to get inserted
+func findInsertionPoint(ei *dosa.EntityInfo, data []map[string]dosa.FieldValue, insertMe map[string]dosa.FieldValue) (found bool, idx int) {
+	found = false
+	idx = sort.Search(len(data), func(offset int) bool {
+		cmp := compareRows(ei, data[offset], insertMe)
+		if cmp == 0 {
+			found = true
+		}
+		return cmp >= 0
+	})
+	return
+}
+
+// compareRows compares two maps of row data based on clustering keys. It handles ascending/descending
+// based on the passed-in schema
+func compareRows(ei *dosa.EntityInfo, v1 map[string]dosa.FieldValue, v2 map[string]dosa.FieldValue) (cmp int8) {
+	keys := ei.Def.Key.ClusteringKeys
+	for _, key := range keys {
+		d1 := v1[key.Name]
+		d2 := v2[key.Name]
+		cmp = compareType(d1, d2)
+		if key.Descending {
+			cmp = -cmp
+		}
+		if cmp != 0 {
+			return cmp
+		}
+	}
+	return cmp
+}
+
+// This function returns the time bits from a UUID
+// You would have to scale this to nanos to make a
+// time.Time but we don't generally need that for
+// comparisons. See RFC 4122 for these bit offsets
+func timeFromUUID(u uuid.UUID) int64 {
+	low := int64(binary.BigEndian.Uint32(u[0:4]))
+	mid := int64(binary.BigEndian.Uint16(u[4:6]))
+	hi := int64((binary.BigEndian.Uint16(u[6:8]) & 0x0fff))
+	return low + (mid << 32) + (hi << 48)
+}
+
+// compareType compares a single DOSA field based on the type. This code assumes the types of each
+// of the columns are the same, or it will panic
+func compareType(d1 dosa.FieldValue, d2 dosa.FieldValue) int8 {
```

Comment:
I'm going to skip these for now. This added quite a few lines of code and didn't seem to buy me anything.